relay/cached_accessor_test.go

```diff
@@ -0,0 +1,248 @@
+
+func TestCacheMisses(t *testing.T) {
+	tu.InitializeRandom()
+
+	cacheSize := rand.Intn(10) + 10
+	keyCount := cacheSize + 1
+
+	baseData := make(map[int]string)
+	for i := 0; i < keyCount; i++ {
+		baseData[i] = tu.RandomString(10)
+	}
+
+	cacheMissCount := atomic.Uint64{}
+
+	accessor := func(key int) (*string, error) {
+		cacheMissCount.Add(1)
+		str := baseData[key]
+		return &str, nil
+	}
+
+	ca, err := NewCachedAccessor(cacheSize, accessor)
+	require.NoError(t, err)
+
+	// Get the first cacheSize keys. This should fill the cache.
+	expectedCacheMissCount := uint64(0)
+	for i := 0; i < cacheSize; i++ {
+		expectedCacheMissCount++
+		value, err := ca.Get(i)
+		require.NoError(t, err)
+		require.Equal(t, baseData[i], *value)
+		require.Equal(t, expectedCacheMissCount, cacheMissCount.Load())
+	}
+
+	// Get the first cacheSize keys again. This should not increase the cache miss count.
+	for i := 0; i < cacheSize; i++ {
+		value, err := ca.Get(i)
+		require.NoError(t, err)
+		require.Equal(t, baseData[i], *value)
+		require.Equal(t, expectedCacheMissCount, cacheMissCount.Load())
+	}
+
+	// Read the last key. This should cause the first key to be evicted.
+	expectedCacheMissCount++
+	value, err := ca.Get(cacheSize)
+	require.NoError(t, err)
+	require.Equal(t, baseData[cacheSize], *value)
+
+	// Read the keys in order. Due to the order of evictions, each read should result in a cache miss.
+	for i := 0; i < cacheSize; i++ {
+		expectedCacheMissCount++
+		value, err := ca.Get(i)
+		require.NoError(t, err)
+		require.Equal(t, baseData[i], *value)
+		require.Equal(t, expectedCacheMissCount, cacheMissCount.Load())
+	}
+}
+
+func TestParallelAccess(t *testing.T) {
+	tu.InitializeRandom()
+
+	dataSize := 1024
+
+	baseData := make(map[int]string)
+	for i := 0; i < dataSize; i++ {
+		baseData[i] = tu.RandomString(10)
+	}
+
+	accessorLock := sync.RWMutex{}
+	cacheMissCount := atomic.Uint64{}
+	accessor := func(key int) (*string, error) {
+
+		// Intentionally block if accessorLock is held by the outside scope.
+		// Used to provoke specific race conditions.
+		accessorLock.Lock()
+		defer accessorLock.Unlock()
+
+		cacheMissCount.Add(1)
+
+		str := baseData[key]
+		return &str, nil
+	}
+	cacheSize := rand.Intn(dataSize) + 1
+
+	ca, err := NewCachedAccessor(cacheSize, accessor)
+	require.NoError(t, err)
+
+	// Lock the accessor. This will cause all cache misses to block.
+	accessorLock.Lock()
+
+	// Start several goroutines that will attempt to access the same key.
+	wg := sync.WaitGroup{}
+	wg.Add(10)
+	for i := 0; i < 10; i++ {
+		go func() {
+			defer wg.Done()
+			value, err := ca.Get(0)
+			require.NoError(t, err)
+			require.Equal(t, baseData[0], *value)
+		}()
+	}
+
+	// Wait for the goroutines to start. We want to give the goroutines a chance to do naughty things if they want.
+	// Eliminating this sleep will not cause the test to fail, but it may cause the test not to exercise the
+	// desired race condition.
+	time.Sleep(100 * time.Millisecond)
```

Comment:
Sleep is generally not reliable in tests. i.e. I don't think it's guaranteed that it always sleep the right amount in test and it can result in flaky tests. 
Maybe we can use something like a buffered channel to wait for all goroutines to trigger?

Reply:
In this scenario, I'm not relying on the sleep to make the test pass. The purpose of the sleep statement is to allow background threads a chance to do bad things if they want to do bad things.

In order to show that test stability does not depend on the sleep, I've set this up so that it runs twice: once with the sleep and once without it. Are you ok with this approach? If you'd still prefer to avoid having a sleep, we should chat. One option might just be to remove the sleep entirely.